apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/DefaultUpload.kt

```diff
@@ -0,0 +1,83 @@
+package com.apollographql.apollo3.api
+
+import okio.BufferedSink
+import okio.BufferedSource
+import okio.ByteString
+import okio.ByteString.Companion.encodeUtf8
+import okio.ByteString.Companion.toByteString
+import okio.use
+
+/**
+ * An [Upload] that writes data from the provided content
+ *
+ * If the content is a [bufferedSource], the [DefaultUpload] will close it once uploaded
+ * If the content is a [byteString], the [DefaultUpload] can be reused
+ */
+class DefaultUpload internal constructor(
+    private val bufferedSource: BufferedSource?,
+    private val byteString: ByteString?,
+    override val contentType: String,
+    override val contentLength: Long,
+    override val fileName: String?,
+) : Upload {
+  override fun writeTo(sink: BufferedSink) {
+    if (bufferedSource != null) {
+      bufferedSource.use {
+        sink.writeAll(it)
+      }
+    } else if (byteString != null) {
+      sink.write(byteString)
+    } else {
+      error("No upload content found")
+    }
+  }
+
+  class Builder {
+    private var bufferedSource: BufferedSource? = null
+    private var byteString: ByteString? = null
+    private var contentType: String? = null
+    private var contentLength: Long = -1
+    private var fileName: String? = null
+
+    fun content(content: BufferedSource): Builder = apply {
+      this.bufferedSource = content
+    }
+
+    fun content(content: String): Builder = apply {
+      this.byteString = content.encodeUtf8()
```

Comment:
To be extra careful, we could disallow setting both `bufferedSource` and `byteString` as this would probably be a mistake (like we do [here](https://github.com/apollographql/apollo-android/blob/f5eb27ade6185544bfa2958bfda6671172ca3b66/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/http/Http.kt#L127) for instance)

Reply:
Good idea 👍 . I took a slightly different approach. Let me know what you think: https://github.com/apollographql/apollo-android/pull/3613/commits/94822de45e3e1e44ec502851603332c54a4761cd

Reply:
Yes, perfect 👍